src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsEntityTypeDefinition.cs

```diff
@@ -112,7 +112,36 @@ private IEnumerable<IEdmStructuralProperty> ComputeDeclaredKey()
                 List<IEdmStructuralProperty> key = new List<IEdmStructuralProperty>();
                 foreach (CsdlPropertyReference keyProperty in this.entity.Key.Properties)
                 {
-                    IEdmStructuralProperty structuralProperty = this.FindProperty(keyProperty.PropertyName) as IEdmStructuralProperty;
+                    IEdmStructuralProperty structuralProperty = null;
+                    if (!string.IsNullOrEmpty(keyProperty?.Alias))
+                    {
+                        string[] keyPropertySegments = keyProperty.PropertyName.Split('/');
+                        if (keyPropertySegments.Length > 0)
+                        {
+                            structuralProperty = this.FindProperty(keyPropertySegments[0]) as IEdmStructuralProperty;
+                            for (int i = 1; i < keyPropertySegments.Length; i++)
+                            {
+                                if (structuralProperty.Type.Definition.TypeKind == EdmTypeKind.Complex)
+                                {
+                                    structuralProperty = structuralProperty.Type.AsComplex().FindProperty(keyPropertySegments[i]) as IEdmStructuralProperty;
+                                }
+                                else if (structuralProperty.Type.Definition.TypeKind == EdmTypeKind.Entity)
+                                {
+                                    structuralProperty = structuralProperty.Type.AsEntity().FindProperty(keyPropertySegments[i]) as IEdmStructuralProperty;
+                                }
+                                else
+                                {
+                                    key.Add(new UnresolvedProperty(this, keyProperty.PropertyName, this.Location));
+                                }
+                            }
+                            structuralProperty = new EdmStructuralPropertyAlias(structuralProperty.DeclaringType, structuralProperty.Name, structuralProperty.Type, keyProperty.Alias, keyPropertySegments);

```

Comment:
Curious, do we perform any validation to verify that the referenced property is not a collection property? For example, if a segment references a collection property `FavoriteColors` of strings that may hold values `["Black","White"]` would that property still be valid to have in the key alias?

Reply:
We check the type of referenced property, if it's not of type complex or entity, then we throw an exception.